### ISSUE TITLE
Redis replies immediately but Lettuce experiences random timeout failures

### ISSUE THREAD

**Issue by Shawyeok** [lettuce-io/lettuce-core#1254]
## Bug Report

We have an application running on production which use lettuce to interact with redis, `del` operation mostly. Recently we found command timeout occurred randomly after startup 1-2 hours, we captured  packets while client command timeout and found redis replay immediate(<1ms), but client get timeout.
![image](https://user-images.githubusercontent.com/5058708/78139861-b5d98980-745b-11ea-8aab-1998cf6f8a60.png)
- packet 431536: DEL request
- packet 431557: response 1
- packet 431559: ack for last packet

#### Current Behavior

```java
command: del, args: [[key**********************************]], spend: 2000ms
```

</details>

#### Input Code

<details>
<summary>Input Code</summary>

```java
// initialize client
ClientResources clientResources = DefaultClientResources.builder()
                .eventLoopGroupProvider(eventLoopGroupProvider) // reuse cross multi client
                .eventExecutorGroup(eventExecutorGroup) // same reused
                .build();
ClusterTopologyRefreshOptions refreshOptions = ClusterTopologyRefreshOptions.builder()
                .enableAdaptiveRefreshTrigger(
                        ClusterTopologyRefreshOptions.RefreshTrigger.MOVED_REDIRECT,
                        ClusterTopologyRefreshOptions.RefreshTrigger.ASK_REDIRECT)
                .build();
RedisClusterClient client = RedisClusterClient.create(resources, uri);
                client.setOptions(ClusterClientOptions.builder()
                        .socketOptions(buildSocketOptions()) // TCP_NODELAY, SO_KEEPALIVE
                        .timeoutOptions(buildTimeoutOptions()) // timeout 2s
                        .topologyRefreshOptions(refreshOptions)
                        .requestQueueSize(2000)
                        .build();
// call del operation
connection.sync().del(key);
```

</details>

#### Environment

- Lettuce version(s): [5.1.2.RELEASE, 5.2.2.RELEASE]
- Netty versions: [4.1.29.Final, 4.1.48.Final]
- Redis version: [4.0.14]

**Comment by mp911de** [comment]
Are you able to reproduce the issue? If so, please capture the state of the request queue and the protocol stack (`CommandHandler.stack`) and the reply buffer (`CommandHandler.buffer`)

**Comment by Shawyeok** [comment]
![image](https://user-images.githubusercontent.com/5058708/78201142-73e42e00-74c3-11ea-8798-afae01e20fa7.png)
sry for delay replay.

we use lz4 codec, code detail: https://gist.github.com/Shawyeok/2395a4cd3de2b43adf37b0f481d81ac3#file-lz4rediscodec-java

**Comment by mp911de** [comment]
Thanks for the detail. Seems like the stack contains items that were not processed and therefore you encounter timeouts. Now we need a way to reproduce the issue to investigate what's going on.

**Comment by Shawyeok** [comment]
Right, the trace log shows that the stack contains 2 commands.
![image](https://user-images.githubusercontent.com/5058708/78319996-07d4f900-759b-11ea-9127-c69b6b8d0682.png)
The column **log_ts** is log time, just ignore the time column, it was delayed.

**Comment by mp911de** [comment]
Note that custom codecs may lead to such a situation when encoding/decoding fails. Please make sure that this isn't the case so that you log encoding exceptions from within your codec.

**Comment by Shawyeok** [comment]
Note that the type of `CommandHandler.stack` is `ArrayDeque`, a non thread-safe deque. Is it  possible lead to concurrent access issues?

**Comment by mp911de** [comment]
All access happens on the Netty I/O thread (offer/poll) to ensure single-threaded access to `ArrayDeque`.

**Comment by Shawyeok** [comment]
Just noticed _requirements_ about `ioThreadPoolSize` and `computationThreadPoolSize` in the lettuce reference guide:

> The minimum I/O threads are 3. A pool with fewer threads can cause undefined behavior.
> The minimum computation threads are 3. A pool with fewer threads can cause undefined behavior.

https://lettuce.io/core/release/reference/index.html#clientresources.using-and-reusing-clientresources

Why magic 3? Why not something else? I'd appreciate if you explain about this.

Coincidentally, my configuration just have 2 threads in `EventLoopGroup` and `EventExecutorGroup` each, may be the cause of this issue?

**Comment by mp911de** [comment]
Previously, we had more blocking operations which caused lockups if there were less than 3 threads available. With Lettuce 6, we should be able to reduce the thread count to 1.

**Comment by Shawyeok** [comment]
Thanks for explanation.

Oops, we did lots of `del` operation in blocking model, but it just block caller thread, still execute in  asynchronous model under the hood, is that right?

By the way, I'll fix misuse about thread size.

**Comment by mp911de** [comment]
Commands are written to the transport channel using a Netty I/O thread. You can issue further commands when your code is called back on netty threads but only if you're using the async API. Using the blocking API when being called on a netty thread will sooner or later lock up the entire application.

Without a reproducer, we're not able to progress on this ticket.

**Comment by Shawyeok** [comment]
Also call async api in production, but mostly call blocking del operation as I said.

I'll try write a case that able to reproduce the problem in weekend.

**Comment by Shawyeok** [comment]
> Note that custom codecs may lead to such a situation when encoding/decoding fails. Please make sure that this isn't the case so that you log encoding exceptions from within your codec.

Problem resolved, you are right in the first place. @mp911de 

We enabled TRACE logging and found `io.netty.handler.codec.EncoderException` in the log. We found a code issue that will cause a `ConcurrentModificationException` in the command encode process, therefor cause a codec encode error. We cannot catch this error normally because just call `setAsync` and didn't wait for result at all. :-(
